### PR TITLE
aerc 0.1.2

### PR DESCRIPTION
**Pull request by notpeter**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----
~~Depends upon `scdoc` formula.  Merge after scdoc is available: https://github.com/Homebrew/homebrew-core/pull/40662.~~ Merged 6/7/19

### CODE REVIEW

**Comment by Moisan** [comment]
@BrewTestBot test this please

**Comment by PeterFaiman** [comment]
New release:

```ruby
url "https://git.sr.ht/~sircmpwn/aerc/archive/0.1.1.tar.gz"
sha256 "eb9f6cf6f77b1eef25d04d7c7cd05fdaebd3d41cd47fa95230b222acac94fc66"
```

**Comment by nnathan** [comment]
you may want to update the title of the PR to: aerc 0.1.1 (new formula); also might want to (somehow) retrigger the CI now that scdoc is merged!

**Comment by SMillerDev** [comment]
Updating to 0.1.1 will trigger the build

**Comment by notpeter** [comment]
Great! I've updated to 0.1.1 and Jenkins is passing. Let me know if there's anything else needed for this to merge.

File: Formula/aerc.rb
Comment (fxcoudert) [inline Formula/aerc.rb]: Why is this necessary? `python3` is in Homebrew PATH, so it should work as it is.

File: Formula/aerc.rb
Comment (fxcoudert) [inline Formula/aerc.rb]: We need a test that exercises the some of the functionality of the app. Version checks or usage checks (`foo --version` or `foo --help`) are not sufficient, as explained in the [formula cookbook](https://docs.brew.sh/Formula-Cookbook).

In most cases, a good test would involve running a simple test case: run `#{bin}/foo input.txt`.
- Then you can check that the output is as expected (with `assert_equal` or `assert_match` on the output of `shell_output`)
- You can also check that an output file was created, if that is expected: `assert_predicate testpath/"output.txt", :exist?`

Some advice for specific cases:
- If the formula is a library, compile and run some simple code that links against it. It could be taken from upstream's documentation / source examples.
- If the formula is for a GUI program, try to find some function that runs as command-line only, like a format conversion, reading or displaying a config file, etc.
- If the software cannot function without credentials, a test could be to try to connect with invalid credentials (or without credentials) and confirm that it fails as expected.
- Same if the software requires a virtual machine, docker instance, etc. to be running.

**Review by fxcoudert — CHANGES_REQUESTED** [review]

File: Formula/aerc.rb
Comment (Akeboshiwind) [reply Formula/aerc.rb]: Unfortunately` aerc` doesn't seem to have any command line only options yet, but it does log to stdout.

Would it be acceptable if the test used a tool like `timeout` to run the command with some invalid credentials and then checked the log to see if it tried to connect?

If so a solution could look something like this:
```bash
touch ~/.config/aerc/aerc.conf ~/.config/aerc/binds.conf
cat << CONF
[Test]
source   = imaps://test%40test.com@test.com
outgoing = smtps+plain://test%40test.com@test.com
default  = INBOX
from     = Test User <test@test.com>
CONF > ~/.config/aerc/accounts.conf
timeout 5s aerc > /tmp/aerc.log
# test that the log looks like expected
```

I'd imagine that the paths would have to be cleaned up to work with brew and I think `coreutils` might be a requirement for `timeout`, but the above works well enough on my machine.

**Comment by fxcoudert** [comment]
Also: the release notes says it's pre-release, implying that it's not stable software. We normally don't accept this in Homebrew, see: https://github.com/Homebrew/brew/blob/master/docs/Acceptable-Formulae.md

In addition, we don't have a way to determine how used it is… First 0.1.0 release was 5 days ago, so I'd suggest we wait a bit to determine notability. Maybe you can maintain it in tap https://docs.brew.sh/How-to-Create-and-Maintain-a-Tap and from there, if it's popular, we can pick it up in core?

**Comment by nnathan** [comment]
> Also: the release notes says it's pre-release, implying that it's not stable software. We normally don't accept this in Homebrew, see: https://github.com/Homebrew/brew/blob/master/docs/Acceptable-Formulae.md

That page only asks that the formula is "stable" and not declared unstable/beta upstream.

From the release page of [aerc](https://git.sr.ht/~sircmpwn/aerc/refs):

> aerc 0.1.0
> This is the initial pre-release of the aerc email client. This is the
work of 16 authors over the course of 2 years, and **represents a
reasonably stable daily driver**. However, many features are missing which
we hope to add to the final release of aerc.

Also that is for 0.1.0. The following release 0.1.1 which this brew is now updated to represents a regular release which doesn't have any such designation implying a stable release.

File: Formula/aerc.rb
Comment (notpeter) [reply Formula/aerc.rb]: My understanding is that it's poor form to install a python dependency (colorama) into the Homebrew global Python environment and so instead I put it in a virtualenv.  These scripts have pip dependencies and by default are directly exec'd by the main golang executable.  Without manipulating the shebang, they run in the global homebrew env which lacks colorama.

Alternatively, I could wrap execution of the main executable with a shell script that first activates the virtualenv before launching the golang bin.  If there are other preferred method(s) please let me know.

File: Formula/aerc.rb
Comment (notpeter) [reply Formula/aerc.rb]: I absolutely understand that `--version` is a terrible test.  The lack of command-line options and complexity of a mail client made me realize there wasn't a quick win for writing a test. Even the above makes coreutils a build-time requirement which feels overkill.

Comment (nnathan) [reply Formula/aerc.rb]: Is the test really essential? I mean it just sounds like a pain in the arse.

**Comment by nirvdrum** [comment]
As an interested observer, is there anything I can to help move this along? I've been quite interested in trying out aerc and was hoping to use Homebrew to simplify the installation. Likewise, if the PR can't be merged on some policy grounds, that'd be helpful to know too.

**Comment by notpeter** [comment]
I have little interest in maintaining a private tap. Perhaps @ddevault would if there were a clear reason for exclusion from homebrew-core.

But I’m unclear the goal post for acceptance: 
 * notoriety
 * tests of build (e.g. upstream `—verison`)
 * functional tests
 * alternative to sed shebang of python venv 

Can anyone provide clear guidance?

**Comment by SMillerDev** [comment]
>I have little interest in maintaining a private tap. Perhaps @ddevault would if there were a clear reason for exclusion from homebrew-core. ...

@notpeter https://github.com/Homebrew/brew/blob/master/docs/Acceptable-Formulae.md are the requirements for software to be acceptable for homebrew-core. 

Beyond that the formula needs a way to check if a formula works so it needs to test the formula. These test need to be meaningful and actually test a code path that people are going to use. There's not that many people that use their tools just to run `tool --version`. So testing that won't improve the user experience.

As for the sed, I don't really know why it's needed, /usr/bin/env should do the same thing. But that would be "requires patches to build" as defined in the rules.

**Comment by nnathan** [comment]
> Beyond that the formula needs a way to check if a formula works so it needs to test the formula.

If this is a requirement this should be specified in the acceptable formulae documentation.

**Comment by SMillerDev** [comment]
> Frankly my take is that we should just put in a test now, this can be improved as aerc introduces more functionality via cli options.
If there's no unified way (`brew test <formula>`) to verify that the software actually works as it's distributed by homebrew (its main function, not its ability to print a version number) I'd rather not ship it at all. @Homebrew/core any other opinions?

> If this is a requirement this should be specified in the acceptable formulae documentation.

While I agree that this could be documented better, the document I referred to indicates "what software is acceptable as a formula", while the test are part of "How do I make an acceptable pull request out of a formula". This is why it's part of the pull request template.

**Comment by nirvdrum** [comment]
I can appreciate the testing requirement; it appears to help keep the existing formulae of high quality. But, some software can't be tested easily standalone. I'm not wild about `curl` and `links` requiring network access to 3rd party resources, as it makes the tests slightly unreliable and creates issues in environments that require an HTTP proxy. Relying on a 3rd party mail server to execute tests strikes me as being even more problematic.

Integration testing with network resources is just hard. To help move this along, I checked some other formulae to see how they handle it. It looks like the tests for [mutt](https://github.com/Homebrew/homebrew-core/blob/49aac0cb636e262d6d7cf9d16e7e016f188d248b/Formula/mutt.rb#L87-L92) are just a step above checking `--version` output. [curlftpfs](https://github.com/Homebrew/homebrew-core/blob/49aac0cb636e262d6d7cf9d16e7e016f188d248b/Formula/curlftpfs.rb) doesn't even attempt to have a test. [sshfs](https://github.com/Homebrew/homebrew-core/blob/49aac0cb636e262d6d7cf9d16e7e016f188d248b/Formula/sshfs.rb#L27-L29) only does a version check. [git-ftp](https://github.com/Homebrew/homebrew-core/blob/49aac0cb636e262d6d7cf9d16e7e016f188d248b/Formula/git-ftp.rb#L49-L51) only checks that `--help` works, which admittedly is at least something people do. [aws-elastickbeanstalk](https://github.com/Homebrew/homebrew-core/blob/49aac0cb636e262d6d7cf9d16e7e016f188d248b/Formula/aws-elasticbeanstalk.rb#L175-L177) only does a version check as well.

I can keep searching, but I've been unable to find a formula that requires a non-HTTP network resource that tests anything other than trivial functionality that indicates the tool was built properly. I can appreciate new contributions don't need to repeat the sins of the past, but I wonder if there is some sort of policy exemption that exists for software that technically can be tested but which requires an unreasonable test environment to do so. Barring that, I think some guidance on how to go about testing `aerc` would be helpful. Unfortunately, I wasn't able to find anything in the existing formulae.

File: Formula/aerc.rb
Comment (MikeMcQuaid) [reply Formula/aerc.rb]: @nnathan Yes, it is essential. It sounds like "a pain in the arse" to you because you don't have experience running a package manager.

Comment (MikeMcQuaid) [reply Formula/aerc.rb]: > Would it be acceptable if the test used a tool like `timeout` to run the command with some invalid credentials and then checked the log to see if it tried to connect?

Yes, this sounds like perhaps the best test we can do considering.

**Comment by MikeMcQuaid** [comment]
> If this is a requirement this should be specified in the acceptable formulae documentation.

The lack of a test is flagged by `brew audit` and Homebrew maintainers. We cannot document/automate everything. If you would like to submit a pull request to suggest an improvement the documentation: feel free.

> I'm not wild about `curl` and `links` requiring network access to 3rd party resources, as it makes the tests slightly unreliable and creates issues in environments that require an HTTP proxy.

The tests exist primarily for our CI environment in which those tests have and continue to be good tests of functionality.

> Barring that, I think some guidance on how to go about testing `aerc` would be helpful.

I've now suggested above verifying it's written expected strings to a log would be acceptable. If it now provides a `--version` or literally any other way of getting it to run and produce output that indicates it's not immediately segfaulting/failing: that might be an acceptable fallback option.

File: Formula/aerc.rb
Comment (ddevault) [reply Formula/aerc.rb]: There are some tests.

    go test ./...

Will be sufficient.

Comment (MikeMcQuaid) [reply Formula/aerc.rb]: No, it will not. These are runtime integration tests when the source code is no longer available. It is a test of our packaging and linking rather than unit tests distributed with the application.

Comment (ddevault) [reply Formula/aerc.rb]: I see. Well, I'll add a `-v` flag momentarily.

Comment (MikeMcQuaid) [reply Formula/aerc.rb]: @ddevault If you're going to add something to the application specially it'd be great if you could make something a little more involved. Ideally something that e.g. spins up the main "GUI" and then immediately tears it down. Verifying that doesn't immediately crash would be an excellent test.

Comment (ddevault) [reply Formula/aerc.rb]: A version flag is broadly useful and easy to add, but a "spin up the frontend" test is a lot less so in both respects. I've got other stuff to do but I can quickly get the version out of the way for you now and unblock your packaging work.

**Comment by nirvdrum** [comment]
@MikeMcQuaid Thanks for chiming in and providing guidance. Likewise, thanks for clarifying what the target test environment is.

File: Formula/aerc.rb
Comment (MikeMcQuaid) [reply Formula/aerc.rb]: @ddevault Sure, that's reasonable. A nice extension to `--version` would be the version string and any other e.g. compile-time conditional values such as the presence/use/versions of dante/python/w3m/scdoc/go. See `curl --version` for a nice example of this.

File: Formula/aerc.rb
Comment (MikeMcQuaid) [reply Formula/aerc.rb]: @notpeter Would it be possible to just install the python resources into `libexec` without a virtualenv and instead set the `PYTHONPATH` accordingly so it can be found? This feels like the responsibility of the upstream software project's packaging so it'd be good to create a feature request/patch for something similar to be added in future.

File: Formula/aerc.rb
Comment (MikeMcQuaid) [inline Formula/aerc.rb]: Why is the `LDFLAGS=` needed? It would be nice if it's possible for this to be removed in a future version.

File: Formula/aerc.rb
Comment (MikeMcQuaid) [inline Formula/aerc.rb]: Add a newline before this line.

File: Formula/aerc.rb
Comment (ddevault) [reply Formula/aerc.rb]: aerc doesn't have any compile-time conditional variables outside of e.g. the installation prefix. Anyway, apply this patch to get `aerc -v`:

https://git.sr.ht/~sircmpwn/aerc/commit/74af57b6c617382ed0e8cd79626cac98ffc2e549.patch

Comment (MikeMcQuaid) [reply Formula/aerc.rb]: @ddevault If the installation prefix is hardcoded and/or used at all by the application: outputting it would be useful if that's at all possible?

Comment (ddevault) [reply Formula/aerc.rb]: It's possible, and of some use, but I have to weigh the complexity add against the value add, and in this case I don't think it adds up.

Comment (MikeMcQuaid) [reply Formula/aerc.rb]: @ddevault Fair enough, thanks anyway.

File: Formula/aerc.rb
Comment (Akeboshiwind) [reply Formula/aerc.rb]: I believe it was required for `scdoc` which was submitted at the same time. I presume the author of this PR assumed it would be necessary for `aerc`, but it shouldn't be.

File: Formula/aerc.rb
Comment (nnathan) [reply Formula/aerc.rb]: > @nnathan Yes, it is essential. It sounds like "a pain in the arse" to you because you don't have experience running a package manager.

Oi, I have written and run my own package manager, it's essentially a rite of passage for any Slackware user :).

**Comment by nirvdrum** [comment]
@notpeter Do you think you'll be able to find time to integrate changes from the PR review? If not, would you be willing to accept a PR to your branch? I'm willing to help out, but don't want to be getting in the way either.

**Comment by notpeter** [comment]
@nirvdrum I'm not available for the next few days.  I've added you as a collaborator to notpeter/homebrew-core. Feel free to commit changes there to update this PR if you have the time.
thanks

**Comment by nirvdrum** [comment]
@notpeter Sorry, I got a bit ill last week and couldn't contribute. I had planned on looking at things this evening, but I see you've made a few changes so I'll pull those down and test locally. Please let me know if there's anything else I can do.

File: Formula/aerc.rb
Comment (notpeter) [reply Formula/aerc.rb]: @MikeMcQuaid Yeah, official upstream stance is basically "install requirements into your global python or activate a appropriate venv first."  It's definitely a little hacky that their go application just execs some bundled python scripts and displays the output inline, but it's not insane, feels actually kinda UNIXy.

Manipulating the PYTHONPATH would require wrapping every execution of the main go bin in a script, no?  That feels similarly.

Comment (MikeMcQuaid) [reply Formula/aerc.rb]: > Manipulating the PYTHONPATH would require wrapping every execution of the main go bin in a script, no? That feels similarly.

That's pretty easy for us to do. Grep for `bin.env_script_all_files`.

**Comment by notpeter** [comment]
This formula builds a functioning `aerc 0.1.2` just as previous revisions of this PR have build 0.1.0 and 0.1.1.

I've updated the formula to reflect some of the requested changes:

* Whitespace
* Spurious LDFLAGS

I have not made changes regarding:

* Creating a venv / pip installing / modifying bundled python scripts to use said venv
  1. Installing packages in the global python is evil 
  2. It feels a little unpythonic (to me) but feels very unixy (execing to a shell script). Not sure what best practices are for similar things (e.g. calling into a ruby script from a python module).
  3. There is no guidance on how best to package polyglot golang/python projects.
* Lack of a test. 
  * A trivial test (`--version` and no SEGFAULT) is infinitely better than no test.
  * A functional test is measurably better than that.
  * If homebrew required a test either trivial (required) or functional (preferred) test I would support that policy, but there is no such policy and @nirvdrum research has shown this has not been applied in the past.
* Upstream addition of command line flags in support of a trivial test as requested above
  * `aerc -v` has been added upstream in response to the request in this thread, but there are issues.
  * There has not yet been a tagged release which includes [that commit](https://git.sr.ht/~sircmpwn/aerc/commit/74af57b6c617382ed0e8cd79626cac98ffc2e549.patch). 0.1.3 may come out in the coming weeks, but that is TBD.
  * aerc requires an interactive terminal (it is a mail client after all). So even when I applied that patch to the 0.1.2 release `brew test aerc` still fails because one of their golang dependencies (gdamore/tcell) immediately panics when it finds a non-interactive terminal.

I am working upstream to see if we can get a non-interactive compatible version flag in a tagged release and will add a test for this formula as that release becomes unavailable. That said, I think it's unfair to require a purely interactive application to also support execution in a non-interactive environment as an unwritten pre-req for inclusion in homebrew.

Unwritten policies cannot be challenged or changed -- it is effectively creates [secret law](https://en.wikipedia.org/wiki/Secret_law) -- and in an open source project it invites bikshedding. There is no project policy and so everyone feel entitled to voice their personal opinion: that's hostile as shit to contributors.

If anyone knows an effort to create standards/policies for contributing to the homebrew project, I would be happy to contribute to that process. But this PR is not the place for that. It took me an hour to get `aerc` functioning on MacOSX + an hour to write this formula.  And many times that to try and get it merged.

**Comment by MikeMcQuaid** [comment]
> * Creating a venv / pip installing / modifying bundled python scripts to use said venv

Did you check out `bin.env_script_all_files` as mentioned?. To be more explicit with my instructions: please use this instead of a virtualenv.

> * but there is no such policy and @nirvdrum research has shown this has not been applied in the past.

Yes, there is such policy. It's written down (https://docs.brew.sh/Formula-Cookbook#add-a-test-to-the-formula) and was linked by @fxcoudert. That we've not been able to enforce the "`--version` tests are bad" universally does not mean it's not a policy. 

To be clear, though, regardless: the policy is what the Homebrew maintainers say the policy is and sometimes we may need to make exceptions (https://docs.brew.sh/Acceptable-Formulae#sometimes-there-are-exceptions)

> * No tagged release yet includes [that commit](https://git.sr.ht/~sircmpwn/aerc/commit/74af57b6c617382ed0e8cd79626cac98ffc2e549.patch). 0.1.3 may come out in the coming weeks, but that is TBD.

We will need to wait for a tagged release (https://docs.brew.sh/Acceptable-Formulae#stable-versions).

> That said, I think it's unfair to require a purely interactive application to also support execution in a non-interactive environment as an unwritten pre-req for inclusion in homebrew.
> Unwritten policies cannot be challenged or changed -- it is effectively creates [secret law](https://en.wikipedia.org/wiki/Secret_law) -- and in an open source project it invites bikshedding.

There is a written project policy that was linked to you above (https://docs.brew.sh/Formula-Cookbook#add-a-test-to-the-formula) which you said is "unfair". You cannot both argue that it's unfair to be consistent with our policies and that we have unwritten policies.

> There is no project policy and so everyone feel entitled to voice their personal opinion: that's hostile as shit to contributors.

Yes, it would be hostile to contributors if there is no policy. There is extensive policy, though, and even the policy explicitly documents that sometimes maintainers may need or choose to make exceptions (https://docs.brew.sh/Acceptable-Formulae#sometimes-there-are-exceptions). There has been no bikeshedding here: all maintainers except me have been consistent with the documented policy and I've tried to find some wiggle room/compromise to help you out and I don't really appreciate having that thrown back in my face.

> If anyone knows an effort to create standards/policies for contributing to the homebrew project, I would be happy to contribute to that process.

There are extensive standards and policies. There is `brew audit`, `brew audit --strict`, `brew style` (run by `brew audit`) and the various documentation on docs.brew.sh.

---

Just so we're clear here: @fxcoudert reviewed your PR within 4 days and pointed out the two unaddressed blockers at that point. One of these is already mentioned in the documentation (`--version` tests) and in `brew create`'s output (`For Homebrew/homebrew-core this will need to be a test that verifies the functionality of the software`) and one I've literally never seen before in working on this project for 10 years (a `go` project that needs vendored `python` modules) so it's unrealistic to expect it be preemptively documented.

This software could have been added to a tap immediately but for it to be in homebrew-core we need to be confident that it can be regularly and trivially updated without breaking. These requirements above all have that root: without them being addressed we are concerned the software will break and, once it's merged, the Homebrew maintainers are on the line for supporting and fixing that breakage and not you.

I understand that it's frustrating to put a bunch of effort in on this software and have it not yet included but that's at least partly a product of the upstream project being difficult to package and test and blockers being unaddressed.

**Comment by PeterFaiman** [comment]
>This software could have been added to a tap immediately but for it to be in homebrew-core we need to be confident that it can be regularly and trivially updated without breaking.

Excellent point. Users of aerc probably already know it's not stable. But putting it in homebrew-core rather than a tap definitely implies the opposite.

Since [tap takes an optional URL argument](https://docs.brew.sh/Taps), this is a good opportunity to put the tap on sr.ht, where the original source is.

**Comment by twe4ked** [comment]
I've pushed up a tap to my GitHub for now.

https://github.com/twe4ked/homebrew-formulae

**Comment by jfdhuiz** [comment]
Your tap only goes until v 0.1.4 while 0.2.1 is released

https://git.sr.ht/~sircmpwn/aerc/refs

The author switched from GitHub to their own Git repo (Sourcehut) at https://git.sr.ht/~sircmpwn/aerc the GH one is read-only, and out of date.

How does this MUA compare with [meli](https://github.com/meli/meli)? Which one is more mature?

**Comment by nirvdrum** [comment]
@jfdhuiz It's probably best to take discussion about a tap to the tap's repo. If you're just looking for a Formula that works with 0.2.1, you can look at my fork of @twe4ked's tap: https://github.com/nirvdrum/homebrew-formulae/blob/master/Formula/aerc.rb. I guess this is a good reminder to open a PR for that.

As for product comparison, that probably shouldn't be on the issue tracker either. I appreciate the convenience of doing so, since presumably you have a captive aerc audience here, but it gums up the issue tracker for Homebrew-related items. On that note, however, I've never heard of meli. It looks interesting. I'd recommend asking more on the [aerc mailing list](https://lists.sr.ht/~sircmpwn/aerc).

**Pull request closed without merging**
